src/relative.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2012-2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Taken from the pathdiff crate, which adapted the original rustc's path_relative_from
+// https://github.com/Manishearth/pathdiff/blob/master/src/lib.rs
+// https://github.com/rust-lang/rust/blob/e1d0de82cc40b666b88d4a6d2c9dcbc81d7ed27f/src/librustc_back/rpath.rs#L116-L158
+
+use std::path::*;
+
+/// Provides helper methods on Path and PathBuf to for creating RelativePath.
+pub trait PathExt {
+    /// Make a relative path from a provided root directory path.
+    ///
+    /// This may not function correctly if either path includes a symlink. Resolve or discard
+    /// symlink paths before making them relative.
+    ///
+    /// ```rust
+    /// use relative_path::PathExt;
+    /// use std::path::*;
+    ///
+    /// let baz = PathBuf::from("/foo/bar/baz");
+    /// let bar = PathBuf::from("/foo/bar");
+    /// let quux = PathBuf::from("/foo/bar/quux");
+    /// assert_eq!(bar.relative_to(&baz), Ok("../".into()));
+    /// assert_eq!(baz.relative_to(&bar), Ok("baz".into()));
+    /// assert_eq!(quux.relative_to(&baz), Ok("../quux".into()));
+    /// assert_eq!(baz.relative_to(&quux), Ok("../baz".into()));
+    /// assert_eq!(bar.relative_to(&quux), Ok("../".into()));
+    ///
+    /// assert_eq!(baz.as_path().relative_to(&bar), Ok("baz".into()));
+    /// assert_eq!(baz.as_path().relative_to(bar.as_path()), Ok("baz".into()));
+    /// ```
+    fn relative_to(&self, root: impl AsRef<Path>) -> Result<RelativePathBuf, FromPathError>;
+}
+
+impl PathExt for Path {
+    fn relative_to(&self, root: impl AsRef<Path>) -> Result<RelativePathBuf, FromPathError> {
+        let root = root.as_ref();
+        if self.is_absolute() != root.is_absolute() {
+            Err(FromPathErrorKind::NonRelative.into())
+        } else {
+            let mut path_comps = self.components();
+            let mut root_comps = root.components();
+            let mut comps: Vec<Component> = vec![];
```

Comment:
This could probably be made directly into a `RelativePathBuf` instead of going through a vector intermediate?

Reply:
I think it'd be better to keep it as a vec for a few reasons. The algorithm is working on a list of std::path::Components. It pops from the 

I prefer a Vec here because the algorithm tends to think about the output as a list of `std::path::Component`. I'd worry it'd become even more incomprehensible if we tried to optimize away the Vec. I did end up cleaning up the conversion from the vec to a relativepathbuf though